engines/ml/xgboost/src/main/java/ai/djl/ml/xgboost/XgbNDArray.java

```diff
@@ -88,14 +88,36 @@ public ByteBuffer toByteBuffer(boolean tryDirect) {
     /** {@inheritDoc} */
     @Override
     public void intern(NDArray replaced) {
-        if (handle != null && handle.get() != 0L) {
-            long pointer = handle.getAndSet(0L);
-            JniUtils.deleteDMatrix(pointer);
+        if (replaced == null) {
+            throw new IllegalArgumentException("The replaced NDArray cannot be null.");
+        }
+        if (!(replaced instanceof XgbNDArray)) {
+            throw new IllegalArgumentException(
+                    "The replaced NDArray must be an instance of XgbNDArray.");
         }
         XgbNDArray array = (XgbNDArray) replaced;
-        data = array.data;
-        handle = array.handle;
-        format = array.format;
+
+        synchronized (this) {
+            if (handle != null && handle.get() != 0L) {
+                long pointer = handle.getAndSet(0L);
+                JniUtils.deleteDMatrix(pointer);
+            }
+
+            data = array.data;
+            format = array.format;
+
+            if (array.handle != null && array.handle.get() != 0L) {
+                if (handle == null) {
+                    handle = new AtomicLong();
+                }
+                handle.set(array.handle.getAndSet(0L));
+            }
+        }
+
+        array.data = null;
+        array.handle = null;
+        array.format = null;
+        array.close();
```

Comment:
The memory leak issue may caused by alternativeArray leak, I think we should close current array's alternative array as will. The code can be re-implemented as the following:

```
        if (!(replaced instanceof XgbNDArray)) {
            throw new IllegalArgumentException(
                    "The replaced NDArray must be an instance of XgbNDArray.");
        }
        XgbNDArray array = (XgbNDArray) replaced;
        if (isReleased()) {
            throw new IllegalArgumentException("This array is already closed");
        }
        if (replaced.isReleased()) {
            throw new IllegalArgumentException("This target array is already closed");
        }

        long pointer = handle.getAndSet(0L);
        JniUtils.deleteDMatrix(pointer);
        if (alternativeArray != null) {
            alternativeArray.close();
        }

        data = array.data;
        handle = array.handle;
        format = array.format;
        alternativeArray = array.alternativeArray;
        array.handle = null;
        array.alternativeArray = null;
        array.close();

```